library/ssl_tls13_generic.c

```diff
@@ -908,13 +908,46 @@ static int ssl_tls13_select_sig_alg_for_certificate_verify(
                                           uint16_t *algorithm )
 {
     uint16_t *sig_alg = ssl->handshake->received_sig_algs;
+    psa_algorithm_t psa_alg = 0;
 
     *algorithm = MBEDTLS_TLS1_3_SIG_NONE;
     for( ; *sig_alg != MBEDTLS_TLS1_3_SIG_NONE ; sig_alg++ )
     {
+        switch( *sig_alg )
+        {
+            case MBEDTLS_TLS1_3_SIG_ECDSA_SECP256R1_SHA256:
+                psa_alg = PSA_ALG_ECDSA( PSA_ALG_SHA_256 );
+                break;
+            case MBEDTLS_TLS1_3_SIG_ECDSA_SECP384R1_SHA384:
+                psa_alg = PSA_ALG_ECDSA( PSA_ALG_SHA_384 );
+                break;
+            case MBEDTLS_TLS1_3_SIG_ECDSA_SECP521R1_SHA512:
+                psa_alg = PSA_ALG_ECDSA( PSA_ALG_SHA_512 );
+                break;
+            case MBEDTLS_TLS1_3_SIG_RSA_PSS_RSAE_SHA256:
+                psa_alg = PSA_ALG_RSA_PSS( PSA_ALG_SHA_256 );
+                break;
+            case MBEDTLS_TLS1_3_SIG_RSA_PSS_RSAE_SHA384:
+                psa_alg = PSA_ALG_RSA_PSS( PSA_ALG_SHA_384 );
+                break;
+            case MBEDTLS_TLS1_3_SIG_RSA_PSS_RSAE_SHA512:
+                psa_alg = PSA_ALG_RSA_PSS( PSA_ALG_SHA_512 );
+                break;
+            case MBEDTLS_TLS1_3_SIG_RSA_PKCS1_SHA256:
+            case MBEDTLS_TLS1_3_SIG_RSA_PKCS1_SHA384:
+            case MBEDTLS_TLS1_3_SIG_RSA_PKCS1_SHA512:
+                psa_alg = PSA_ALG_RSA_PKCS1V15_CRYPT;
+                break;
+            default:
+                break;
+        }
+
         if( mbedtls_ssl_sig_alg_is_offered( ssl, *sig_alg ) &&
             mbedtls_ssl_tls13_sig_alg_for_cert_verify_is_supported( *sig_alg ) &&
-            mbedtls_ssl_tls13_check_sig_alg_cert_key_match( *sig_alg, own_key ) )
+            mbedtls_ssl_tls13_check_sig_alg_cert_key_match( *sig_alg, own_key ) &&
```

Comment:
I think `mbedtls_ssl_tls13_check_sig_alg_cert_key_match` and `mbedtls_pk_can_do_ext` try to achieve the same: check that we will be able to compute the signature with signature algorithm `*sig_alg`/`psa_alg` and private key `own_key`. We should thus probably just remove the call to `mbedtls_ssl_tls13_check_sig_alg_cert_key_match` here. One thing I am not sure about though. In `mbedtls_ssl_tls13_check_sig_alg_cert_key_match` we check the key size and I do not see checks of the key size in `mbedtls_pk_can_do_ext`. @mpg Are those key size checks not necessary ?

Reply:
What should we do with this one?
At the moment for me both checks are needed. `mbedtls_ssl_tls13_check_sig_alg_cert_key_match` checks key sizes and `mbedtls_pk_can_do_ext` verifies opaque key usage flags.

Reply:
I've discussed that with Manuel offline and I will rather try to get rid completely of `mbedtls_pk_can_do_ext` call by trying different plausible algorithms in `ssl_tls13_write_certificate_verify_body`. By plausible I mean that it is proposed by the peer, supported by the sender and that it is compatible with the key type (no point in trying to do an ECDSA signature if we have an RSA key). That way for TLS 1.3 (I know it is not possible for TLS 1.2) we will not rely on the duplication of the complicated checks done by the PSA crypto implementation to know upfront if a signature will succeed or not. There is a time cost in trying different signature algorithm but in most cases the first try will be the good one I think.

Reply:
In practice, this means we should pause this PR until we know if Ronald's approach works. If it does, then it will supersede this PR (as the approach seems better than the one I had suggested), and if not, we can resume this PR.

Reply:
Sorry for the inconvenience (starting then pausing and potentially dropping this), I didn't realise before chatting with Ronald that for TLS 1.3 the solution space was larger than for TLS 1.2, and that we had a chance to avoid using `can_do_ext()` at all.